src/main/java/org/testng/internal/Invoker.java

```diff
@@ -579,6 +580,29 @@ private ITestResult invokeMethod(
       ITestNGMethod[] afterMethods,
       ConfigurationGroupMethods groupMethods,
       FailureContext failureContext) {
+
+    if(System.getProperties().containsKey("testng.newInstancePerMethod") ||
+       System.getenv().containsKey("TESTNG_NEW_INSTANCE_PER_METHOD") ||
+       instance.getClass().isAnnotationPresent(NewInstancePerMethod.class)) {
+      boolean tryConstructor = true;
+      if (Cloneable.class.isAssignableFrom(instance.getClass())) {
+        try {
+          instance = instance.getClass().getDeclaredMethod("clone").invoke(instance);
+          tryConstructor = false;
+        } catch (NoSuchMethodException e) {
+        } catch (IllegalAccessException e) {
+        } catch (InvocationTargetException e) {
+        }
+      }
+      if(tryConstructor) {
+        try {
+          instance = ClassHelper.newInstance(instance.getClass());
```

Comment:
It looks that this will make the TestNG code messy. There are multiple internal methods which uses `Object instance` passed as argument. So the behavior will be different for different methods for using instance of class. For example, method `invokeConfigurationMethod`! Either this should be applied in such way that each methods uses same mechanism or should be handled outside TestNG code.